scripts/word_embeddings/tools/cooccur.cc

```diff
@@ -0,0 +1,369 @@
+    if (size() < capacity()) {
+      _data.push_back(item);
+    } else {
+      _data.at(_pointer) = item;
+      _pointer = (_pointer + 1) % _capacity;
+    }
+  }
+
+  void clear() {
+    _data.clear();
+    _pointer = 0;
+  }
+
+private:
+  const size_t _capacity;
+  std::vector<T> _data;
+  size_t _pointer = 0;
+
+  // Ranges TS based iterator over circular buffer
+  friend ranges::range_access;
+  struct cursor {
+    cursor() = default;
+    cursor(size_t pointer, size_t counter, const CircularBuffer *parent)
+        : pointer(pointer), counter(counter), parent(parent) {}
+
+    T read() const { return parent->_data.at(pointer); }
+    bool equal(const cursor &other) const { return counter == other.counter; }
+    void next() {
+      // LIFO
+      if (pointer == 0) {
+        pointer = parent->size() - 1;
+      } else {
+        pointer -= 1;
+      }
+      counter++;
+    }
+
+    size_t pointer;
+    size_t counter;
+    const CircularBuffer *parent;
+  };
+  cursor begin_cursor() const {
+    if (size() == 0)
+      return cursor(_pointer, 0, this);
+    else if (_pointer == 0)
+      return cursor(size() - 1, 0, this);
+    else
+      return cursor(_pointer - 1, 0, this);
+  }
+  cursor end_cursor() const { return cursor(_pointer, size(), this); }
+};
+
+// * Input
+auto ReadVocab() {
+  std::string word;
+  std::string count;
+  int rank{0};
+  Vocab vocab;
+  while (std::cin >> word) {
+    std::cin >> count;
+    vocab[word] = {rank, std::stoi(count)};
+    rank++;
+  }
+  return vocab;
+}
+// * Co-occurence matrix construction
+std::mutex paths_m;
+std::mutex matrices_m;
+
+void ReadMatrix(std::queue<fs::path> &paths, queue<Matrix> &matrices,
+                const Vocab &vocab, const Arguments &args, uint32_t seed) {
+  assert(seed > 0);
+  std::string line;
+  CircularBuffer<uint32_t> history(args.window_size);
+  std::unique_ptr<Matrix> m = std::make_unique<Matrix>();
+
+  // Prepare subsampling
+  std::random_device r;
+  std::default_random_engine random_engine(r());
+  std::uniform_real_distribution<float> uniform_dist(0, 1);
+  std::vector<double> idx_to_pdiscard;
+  if (args.subsample) {
+    auto counts = vocab | rsv::values | rsv::values;
+    double sum_counts = std::accumulate(counts.begin(), counts.end(), 0);
+    double t = 1E-4;
+    for (const double count : vocab | rsv::values | rsv::values) {
+      idx_to_pdiscard.push_back(1 - std::sqrt(t / (count / sum_counts)));
+    }
+  }
+
+  while (true) {
+    fs::path path;
+    {
+      std::scoped_lock lock(paths_m);
+      if (paths.empty()) {
+        break;
+      }
+      path = paths.front();
+      paths.pop();
+    }
+
+    std::ifstream in{path};
+    if (!in.is_open()) {
+      throw std::invalid_argument(path.string() + " cannot be opened!");
+    }
+    while (std::getline(in, line)) {
+      history.clear(); // Discard context from other lines
+
+      auto words = line | rsv::split((int (*)(int)) & std::isspace);
+      for (auto word : words) {
```

Comment:
also, use `(const string & word : words)` to reduce a copy (and make it read only)

Reply:
I'm not sure if this helps, in C++11 and later `std::string` is guaranteed to end with a `\0`, so by using `std::string` for every token here, we always force a copy of the tokens. C++17 introduces std::string_view as a workaround, but until https://wg21.link/P0919 is merged for upcoming C++20 we can't use `std::string_view` as a key in unordered containers that maps from `std::string` to X.
That's my understanding, but please let me know if you know anything else

Reply:
Does it still create a copy with a reference `&`?